### PR TITLE
configuration-mapping-fields component creation

### PR DESCRIPTION
**Pull request by DhaaraniCIT**
## Summary by CodeRabbit

- **New Features**
	- Introduced a dynamic form for configuring mapping fields for expenses, including dropdown menus, validation, and the ability to add/remove fields dynamically.

- **Enhancements**
	- Improved modularity by importing `SharedModule` in `travelperk-shared.module`.
	- Added `ConfigurationMappingFieldsComponent` to `SharedModule` for broader application use.

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
@coderabbitai review

**Comment by ashwin1111** [comment]
@coderabbitai review

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() SourceFields: string[];
+
+  @Input() DestinationFields: string[];
+
+  @Input() isSourceDisabled: boolean;
+
+  @Input() isDestinationDisabled: boolean;
+
+  @Input() noOfFieldsToShow: number = 0;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:30]: no need of this field

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() showCrossIcon: boolean;
+
+  @Input() appName: string;
+
+  currentLength: number = this.noOfFieldsToShow;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:36]: not needed

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  templateUrl: './configuration-mapping-fields.component.html',
+  styleUrls: ['./configuration-mapping-fields.component.scss']
+})
+export class ConfigurationMappingFieldsComponent implements OnInit {
+
+  @Input() form: FormGroup;
+
+  @Input() sourceFieldText: string;
+
+  @Input() DestinationFieldText: string;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:16]: d small

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() sourceFieldPlaceholderText: string;
+
+  @Input() DestinationPlaceholderFieldText: string;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:20]: d small

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() SourceFields: string[];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:22]: sourceFieldOptions

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() DestinationFields: string[];
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:24]: destinationFieldOptions

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Input() isSourceDisabled: boolean;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:26]: source won't be disabled at any point of time, can remove this

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  removeFilter(expenseField: AbstractControl) {
+    (expenseField as FormGroup).controls.source_field.patchValue('');
+    (expenseField as FormGroup).controls.import_to_fyle.patchValue(false);
+    (expenseField as FormGroup).controls.import_to_fyle.enable();
+    event?.stopPropagation();
+  }
+
+  onDropdownChange(event: any, index: number) {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:68]: remove whole function, not required

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  hasDuplicateOption(formGroup: AbstractControl, index: number, controlName: string): boolean {
+    return (formGroup as FormGroup).controls[controlName].valid;
+  }
+
+  removeFilter(expenseField: AbstractControl) {
+    (expenseField as FormGroup).controls.source_field.patchValue('');
+    (expenseField as FormGroup).controls.import_to_fyle.patchValue(false);
+    (expenseField as FormGroup).controls.import_to_fyle.enable();
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:64]: not required

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+                                <div class="tw-flex tw-justify-between tw-items-center">
+                                    <span>
+                                        {{ item.display_name | snakeCaseToSpaceCase | titlecase }}
+                                    </span>
+                                </div>
+                            </ng-template>
+                        </p-dropdown>
+                        <p *ngIf="expenseField.value.destination_field && !expenseField.valid && !hasDuplicateOption(expenseField, i, 'destination_field')" class="tw-text-alert-toast tw-text-12-px tw-pt-4-px">
+                            {{appName}} Fields should be unique
+                        </p>
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html:27]: please remove these, these are not relevant

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                            <ng-template pTemplate="selectedItem" let-item>
+                            <div class="tw-flex tw-justify-between tw-items-center ">
+                                <span>
+                                    {{ item.display_name }}
+                                </span>
+                                <i class="pi pi-times" (click)="removeFilter(expenseField)"></i>
+                            </div>
+                            </ng-template>
+                      </p-dropdown>
+                      <p *ngIf="expenseField.value.source_field && !expenseField.valid && !hasDuplicateOption(expenseField, i, 'source_field')" class="tw-text-alert-toast tw-text-12-px tw-pt-4-px">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html:52]: same

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+      return true;
+    }
+    return false;
+  }
+
+  addExpenseFields() {
+    this.currentLength = this.form.controls.expenseFields.value.length;
+  }
+
+  hasDuplicateOption(formGroup: AbstractControl, index: number, controlName: string): boolean {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:57]: remove

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+  showOrHideLoadMoreButton() {
+    if (this.form.controls.expenseFields.value.length > this.currentLength) {
+      return true;
+    }
+    return false;
+  }
+
+  addExpenseFields() {
+    this.currentLength = this.form.controls.expenseFields.value.length;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:54]: emit output to parent and call API from there

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  showOrHideLoadMoreButton() {
+    if (this.form.controls.expenseFields.value.length > this.currentLength) {
+      return true;
+    }
+    return false;
+  }
+
+  addExpenseFields() {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:53]: rename func better

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  readonly brandingConfig = brandingConfig;
+
+  constructor() { }
+
+  get expenseFieldsGetter() {
+    return this.form.get('expenseFields') as FormArray;
+  }
+
+  showOrHideLoadMoreButton() {
+    if (this.form.controls.expenseFields.value.length > this.currentLength) {
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:47]: condition should change, receive input from parent on totalAvailableRows

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  @Input() appName: string;
+
+  currentLength: number = this.noOfFieldsToShow;
+
+  readonly brandingConfig = brandingConfig;
+
+  constructor() { }
+
+  get expenseFieldsGetter() {
+    return this.form.get('expenseFields') as FormArray;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.ts:43]: let's call them paymentProfileMappings

File: src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html [hunk 17]
```diff
@@ -0,0 +1,65 @@
+<div class="tw-p-24-px">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-mapping-fields/configuration-mapping-fields.component.html:1]: will review template once ts comments are fixed

**Comment by DhaaraniCIT** [comment]
fixed the comments in https://github.com/fylein/fyle-integrations-app/pull/508

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**